api_app/tests_ma/test_services/test_cost_service.py

```diff
@@ -294,3 +272,230 @@ def __set_shared_service_repo_mock_return_value_without_display_name(shared_serv
+    assert workspace_cost_report.workspace_services[0].user_resources[1].name == "VM2"
+    assert len(workspace_cost_report.workspace_services[0].user_resources[1].costs) == 3
+    assert workspace_cost_report.workspace_services[0].user_resources[1].costs[0].cost == 164.8
+
+    assert workspace_cost_report.workspace_services[1].id == "9ad6e5d8-0bef-4b9f-91d6-ae33884883a1"
+    assert workspace_cost_report.workspace_services[1].name == "Azure ML"
+    assert len(workspace_cost_report.workspace_services[1].costs) == 3
+    assert workspace_cost_report.workspace_services[1].costs[0].cost == 24.8
+    assert len(workspace_cost_report.workspace_services[1].user_resources) == 2
+    assert workspace_cost_report.workspace_services[1].user_resources[0].id == "6ede6dc0-a1e1-40bd-92d7-3b3adcbec66d"
+    assert workspace_cost_report.workspace_services[1].user_resources[0].name == "Compute Instance 1"
+    assert len(workspace_cost_report.workspace_services[1].user_resources[0].costs) == 3
+    assert workspace_cost_report.workspace_services[1].user_resources[0].costs[0].cost == 164.8
+    assert workspace_cost_report.workspace_services[1].user_resources[1].id == "915760d8-cf09-4cdb-b73b-815e6bfaef6f"
+    assert workspace_cost_report.workspace_services[1].user_resources[1].name == "Compute Instance 2"
+    assert len(workspace_cost_report.workspace_services[1].user_resources[1].costs) == 3
+    assert workspace_cost_report.workspace_services[1].user_resources[1].costs[0].cost == 168.8
+
+
+def __get_cost_management_query_result():
+    query_result = QueryResult()
+    query_result.rows = [
+        [37.6, '"tre_core_service_id":"guy22"', 'USD'],
+        [44.5, '"tre_id":"guy22"', 'USD'],
+        [6.8, '"tre_shared_service_id":"848e8eb5-0df6-4d0f-9162-afd9a3fa0631"', 'USD'],
+        [4.8, '"tre_shared_service_id":"f16d0324-9027-4448-b69b-2d48d925e6c0"', 'USD'],
+        [1.8, '"tre_workspace_id":"19b7ce24-aa35-438c-adf6-37e6762911a6"', 'USD'],
+        [2.8, '"tre_workspace_id":"d680d6b7-d1d9-411c-9101-0793da980c81"', 'USD'],
+        [5.8, '"tre_workspace_id":"d680d6b7-d1d9-411c-9101-0793da980c81"', 'ILS'],
+        [6.6, '"tre_workspace_service_id":"f8cac589-c497-4896-9fac-58e65685a20c"', 'USD'],
+        [9.3, '"tre_workspace_service_id":"9ad6e5d8-0bef-4b9f-91d6-ae33884883a1"', 'USD'],
+        [1.3, '"tre_user_resource_id":"09ed3e6e-fee5-41d0-937e-89644575e78c"', 'USD'],
+        [2.3, '"tre_user_resource_id":"8ce4a294-95ae-45a9-8d48-6525ce84eb5a"', 'USD'],
+        [5.2, '"tre_user_resource_id":"6ede6dc0-a1e1-40bd-92d7-3b3adcbec66d"', 'USD'],
+        [4.1, '"tre_user_resource_id":"915760d8-cf09-4cdb-b73b-815e6bfaef6f"', 'USD'],
+    ]
+    return query_result
+
+
+def __set_cost_management_client_mock_query_result():
+    query_result = QueryResult()
+    query_result.rows = [
+        [31.6, 20220501, '"tre_core_service_id":"guy22"', 'USD'],
```

Comment:
Does it really have to be that complex? Since the actual logic for querying it and mapping it is fairly generic?

Reply:
resolved in person

Reply:
as discussed over teams, the api just query cost management api and extract the values into a TRE hierarchal resource tree, it doesn't aggregate or calculate anything. if a cost record was not generated to a resource in Azure the api will return an empty array for that TRE resource.